content/docs/learning-guide/tutorials/custom-tools/shape-example-cpp.md

```diff
@@ -0,0 +1,383 @@
+
+```cpp
+void ShapeExampleWidget::OnNameInputTextChanged(const QString& text)
+{   
+    m_addShapeNameSuffix->setEnabled(!text.isEmpty());
+}
+```
+
+
+## Buttons
+
+In this step, you will create a collection of buttons that create an entity of different shape types, such as a box, sphere, cone, and so on. 
+
+Again, wrap these UI elements in their own sub-widget, set the layout, and add it to the main widget. 
+
+1. To start, instantiate a `QWidget` called `shapeButtons` and a `QGridLayout` called `gridLayout`.
+   
+2. Later, you will query types of shape components, add buttons to this widget, and add signal listeners to the buttons.
+
+3. Finally, set the layout of `shapeButtons` to `gridLayout`, and add `shapeButtons` to `mainLayout`. 
+
+```cpp
+// 1
+QWidget* shapeButtons = new QWidget(this);
+QGridLayout* gridLayout = new QGridLayout();
+
+// 2
+
+// 3
+shapeButtons->setLayout(gridLayout);
+mainLayout->addWidget(shapeButtons);
+```
+
+### Query types of Shape components
+
+First, query the types of shape components by specifying a list of provided services. In O3DE, all components have a list of provided services, that you can query from. In this example, Shape components provide the "ShapeService". 
+
+1. Create a `AzToolsFramework::EntityCompositionRequests::ComponentServicesList` called `providedServices` and add "ShapeService". 
+
+2. Get a list of component types. Call `AzToolsFramework::EditorComponentAPIBus::BroadcastResult(...)` and dispatch `AzToolsFramework::EditorComponentAPIRequests::FindComponentTypeIdsByService`, which takes a list of services to include (`providedServices`) and services to exclude (an empty list) and stores the component types to `typeIds`.
+
+```cpp
+AzToolsFramework::EntityCompositionRequests::ComponentServicesList providedServices;
+providedServices.push_back(AZ_CRC_CE("ShapeService"));
+
+AZ::ComponentTypeList typeIds;
+AzToolsFramework::EditorComponentAPIBus::BroadcastResult(
+    typeIds,
+    &AzToolsFramework::EditorComponentAPIRequests::FindComponentTypeIdsByService,
+    providedServices,
+    AzToolsFramework::EntityCompositionRequests::ComponentServicesList()
+);
+```
+
+
+Next, get the names of the components to add to the buttons. 
+
+1. Get a list of component names. Call `AzToolsFramework::EditorComponentAPIBus::BroadcastResult(...)` and dispatch `AzToolsFramework::EditorComponentAPIRequests::FindComponentTypeNames`, which takes the list of component types (`typeIds`) and stores the associated of component names to `componentNames`.
+
+```cpp
+AZStd::vector<AZStd::string> componentNames;
+AzToolsFramework::EditorComponentAPIBus::BroadcastResult(
+    componentNames,
+    &AzToolsFramework::EditorComponentAPIRequests::FindComponentTypeNames,
+    typeIds
+);
+```
+
+{{ note }}
+You can query a list of services that components provide by calling the component's `GetProvidedServices()` method.
+{{ /note }}
+
+### Add buttons
+
```

Comment:
I would love to see screenshots under this headings of what the heading is referring to. E.g. Buttons